panther_battery/src/adc_node.cpp

```diff
@@ -0,0 +1,191 @@
+  const auto battery_current_window_len = this->get_parameter("batery_current_window_len").as_int();
+  const auto battery_charge_window_len = this->get_parameter("batery_charge_window_len").as_int();
+
+  adc0_reader_ = std::make_unique<ADCDataReader>(adc0_device_);
+  adc1_reader_ = std::make_unique<ADCDataReader>(adc1_device_);
+  last_battery_info_time_ = rclcpp::Time(int64_t(0), RCL_ROS_TIME);
+
+  battery_count_ = CheckBatteryCount();
+
+  // create battery publisher based on battery count
+  if (battery_count_ == 2) {
+    battery_pub_ = std::make_unique<BatteryPublisher>(
+      this->create_publisher<BatteryStateMsg>("battery", 10), high_bat_temp_,
+      2 * bat_charging_curr_thresh_, battery_voltage_window_len, battery_temp_window_len,
+      battery_current_window_len, battery_charge_window_len, 2 * bat_designed_capacity_);
+    battery_1_pub_ = std::make_unique<BatteryPublisher>(
+      this->create_publisher<BatteryStateMsg>("battery_1", 10), high_bat_temp_,
+      bat_charging_curr_thresh_, battery_voltage_window_len, battery_temp_window_len,
+      battery_current_window_len, battery_charge_window_len, bat_designed_capacity_);
+    battery_2_pub_ = std::make_unique<BatteryPublisher>(
+      this->create_publisher<BatteryStateMsg>("battery_2", 10), high_bat_temp_,
+      bat_charging_curr_thresh_, battery_voltage_window_len, battery_temp_window_len,
+      battery_current_window_len, battery_charge_window_len, bat_designed_capacity_);
+  } else {
+    battery_pub_ = std::make_unique<BatteryPublisher>(
+      this->create_publisher<BatteryStateMsg>("battery", 10), high_bat_temp_,
+      bat_charging_curr_thresh_, battery_voltage_window_len, battery_temp_window_len,
+      battery_current_window_len, battery_charge_window_len, bat_designed_capacity_);
+  }
+
+  io_state_sub_ = this->create_subscription<IOStateMsg>(
+    "hardware/io_state", 10, std::bind(&ADCNode::IOStateCB, this, _1));
+
+  // running at 10 Hz
+  battery_pub_timer_ = this->create_wall_timer(
+    std::chrono::milliseconds(100), std::bind(&ADCNode::BatteryPubTimerCB, this));
+
+  RCLCPP_INFO(this->get_logger(), "Battery count: %d", battery_count_);
+  RCLCPP_INFO(this->get_logger(), "Node started");
+}
+
+int ADCNode::CheckBatteryCount()
+{
+  auto trials_num = 10;
+  auto V_temp_sum = 0.0;
+  auto V_temp_bat_2 = 0.0;
+
+  try {
+    for (int i = 0; i < trials_num; i++) {
+      V_temp_sum += adc0_reader_->GetADCMeasurement("in_voltage0_raw", 0.0, 0.002);
+      std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    }
+
+    V_temp_bat_2 = V_temp_sum / trials_num;
+
+  } catch (const std::runtime_error & e) {
+    RCLCPP_ERROR(
+      this->get_logger(),
+      "Error reading ADC data: %s. "
+      "\nThe number of batteries cannot be determined. The single battery was adopted",
+      e.what());
+    return 1;
+  }
+
+  return V_temp_bat_2 > bat02_detect_thresh_ ? 1 : 2;
```

Comment:
```suggestion
   const float V_temp_bat_2 = V_temp_sum / trials_num;
  return V_temp_bat_2 > bat02_detect_thresh_ ? 1 : 2;
```
This will make everything more compact

Reply:
Done